test/unit/master/test_pagination.py

```diff
@@ -0,0 +1,118 @@
+from unittest.mock import MagicMock, Mock
+
+from app.master.build import Build
+from app.master.build_request import BuildRequest
+from app.master.cluster_master import ClusterMaster
+from app.slave.cluster_slave import SlaveState
+from app.util.exceptions import BadRequestError, ItemNotFoundError
+from app.web_framework.cluster_base_handler import ClusterBaseHandler, pagination_constants
+from test.framework.base_unit_test_case import BaseUnitTestCase
+
+
+@genty
+class TestClusterMaster(BaseUnitTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.patch('app.util.fs.create_dir')
+        self.patch('app.util.fs.async_delete')
+        self.patch('os.makedirs')
+        self.mock_slave_allocator = self.patch('app.master.cluster_master.SlaveAllocator').return_value
+        self.mock_scheduler_pool = self.patch('app.master.cluster_master.BuildSchedulerPool').return_value
+
+    @genty_dataset(
+        no_params=(
+            None, None,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + pagination_constants['DEFAULT_LIMIT']
+        ),
+        offset_param=(
+            50, None,
+            50 + 1,
+            50 + pagination_constants['DEFAULT_LIMIT']
+        ),
+        limit_param=(
+            None, 100,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + 100
+        ),
+        offset_and_limit_params=(
+            53, 100,
+            53 + 1,
+            53 + 100
+        ),
+        low_offset=(
+            None, 2,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + 2
+        ),
+        too_high_offset=(
+            1000, 100,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + 100
+        ),
+        too_high_limit=(
+            None, 1000,
+            pagination_constants['DEFAULT_OFFSET'] + 1,
+            pagination_constants['DEFAULT_OFFSET'] + pagination_constants['MAX_LIMIT']
+        ),
+        negative_offset=(
+            -50, None,
+            0 + 1,
+            pagination_constants['DEFAULT_OFFSET'] + pagination_constants['DEFAULT_LIMIT']
+        ),
+        negative_limit=(
+            None, -50,
+            None,
+            None
+        ),
+    )
+    def test_pagination_request_with_query_params(
+            self,
+            offset: Optional[int],
+            limit: Optional[int],
+            expected_first_build_id: int,
+            expected_last_build_id: int,
+            ):
+        master = ClusterMaster()
+        for build_id in range(1, 501):
+            build_mock = Mock(spec=Build)
+            build_mock.build_id = build_id
+            master._all_builds_by_id[build_id] = build_mock
+
+        offset, limit = ClusterBaseHandler._validate_arguments(offset, limit)
```

Comment:
Ah I see, so each test file is dedicated to testing a certain file in clusterrunner rather than a test file will be dedicated to testing a certain concept in clusterrunner? That's why I was creating new files for these pagination tests, I didn't know that's how we organized things. 

If that's the case, then yeah it'd make more sense to move these `builds` tests to `unit/master/test_cluster_master` and the pagination tests to `unit/web_framework/test_cluster_base_handler`

The reason I used `_validate_arguments` here is because the arguments that get passed into the `builds` endpoint always go through this function first. I think it's fine and probably makes more sense to remove this function from this test and just give "validated" entries to the tests instead.